### DISCUSSION TITLE
[OpenMeteo] Flatbuffers support?

### DISCUSSION THREAD

**Discussion by 320kg** [breezy-weather/breezy-weather#1280] [Ideas]
### Describe your idea

Ey! Can we support [FlatBuffers](https://github.com/open-meteo/sdk) for OpenMeteo? This will reduce battery life, reduce roaming cost and also make it a tiiiiiiny bit more accurate. Should this be an issue instead? Thx.

### Acknowledgements

- [X] I have searched the [existing issues](https://github.com/breezy-weather/breezy-weather/issues) and [existing discussions](https://github.com/breezy-weather/breezy-weather/discussions/categories/ideas) and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue or discussion.
- [X] The idea I’m suggesting is not in the [list of features that will not be implemented](https://github.com/breezy-weather/breezy-weather/blob/main/README.md#features-that-will-not-be-implemented).
- [X] I have updated the app to **[version 5.2.7 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**. Even if I’m getting updates from F-Droid default repo, I don’t need to uninstall/reinstall to update.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance). I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
My hot take on this: From what I can read, this is mostly useful for historical weather. Refreshing weather on a 10+ years old device happens in the blink of an eye; it doesn't particularly consume battery (~1% on each charge on my side) or data (with the default 1.5 h refresh rate, I barely get to 5 MB of data per month for a location and I use 4 combined sources, not just Open-Meteo which has the most optimized json compared to others). I don't see anything mentioned regarding accuracy. Additionally, this would require adding a specific library just for 1 of the many sources we support... while it works perfectly if we just keep it the Json way.

**Reply by 320kg** [reply]
Well, we could make an "open to contributers" issue for someone who would think this is fun to implement.
5mb per month doesn't sound but it could be kilobytes!

Regarding accuracy, just sending binary floats is more accurate than describing it in a human readable decimal first (eg. 1078.675) and then finding the nearest float again. Usually not by a lot by human standards, just for perfectionists.

**Reply by 320kg** [reply]
Well, we could make an "open to contributers" issue for someone who would think this is fun to implement.
5mb per month doesn't sound like much but it could be kilobytes!

Regarding accuracy, just sending binary floats is a lot more accurate than describing it in a human readable decimal first (eg. 1078.675) and then finding the nearest float again.

**Reply by papjul** [reply]
> 5mb per month doesn't sound but it could be kilobytes!

I just did a test, and an usual request of 13,58 KB (67,18 KB unzipped) is down to 139 B (37,8 KB unzipped), so that's really impressive indeed!

> Regarding accuracy, just sending binary floats is a lot more accurate than describing it in a human readable decimal first (eg. 1078.675) and then finding the nearest float again.

I don't think it would change anything in the app, we don't have this level of granularity shown anywhere, right?

> we could make an "open to contributers" issue for someone who would think this is fun to implement

There are a few downsides:
- you have to add 3 dependencies for everyone, even people not using Open-Meteo, while you can do the same without this library
- I have to maintain "two systems" while I could do with just one
- we lose the human-readable form of the JSON which is extremely useful when troubleshooting (for example, a malformed JSON is easier to detect than a malformed binary file, or just checking the data sent by the server to compare with the actual displayed values in the app when someone claims "the app is showing incorrect data")

So I'm mostly worried about the maintainability, because having external contributors doing a task is one thing, but ensuring it works and troubleshooting it in the long run is another.

So yes, anyone can have fun implementing it on their fork, but I would not tag it as "Open to contributors" because at this point, I don't have the intention of merging it in the main project. However, I would be curious to take a look at an implementation if someone ever does one. A global option on the Open-Meteo source to switch between FlatBuffers and JSON could be helpful to make sure we can switch back any time, especially the first few months it is implemented.

**Reply by 320kg** [reply]
I never said we should remove the old implementation lol

**Reply by papjul** [reply]
I think it should. If FlatBuffers is better and proved to be working as good as JSON, there is no point keeping two algorithms, it just clunkers the code, it's like supporting both JSON and XML versions of a same API, it's not useful. It could be kept on debug version only if we have real issues troubleshooting FlatBuffers.

**Reply by 320kg** [reply]
It was literally just discussed why it should be kept: easily reading malformed responses

EDIT: could be a slowly deprecated thing

**Reply by papjul** [reply]
If the FlatBuffers endpoint gives you a malformed content, it doesn't mean the JSON endpoint will do the same. It's two different formats.

We could keep the JSON algorithm for a period, but after it is confirmed FlatBuffers works well, the JSON algorithm would have to be removed, otherwise it would require a double maintenance. It could be kept in the debug version for developers understanding.
